runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -479,7 +479,32 @@ public void processWatermark1(Watermark mark) throws Exception {
 
   @Override
   public void processWatermark2(Watermark mark) throws Exception {
-    // ignore watermarks from the side-input input
+    if (mark.getTimestamp() == BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()) {
+      // this means we will never see any more side input
+      pushbackDoFnRunner.startBundle();
+
+      BagState<WindowedValue<InputT>> pushedBack =
+          pushbackStateInternals.state(StateNamespaces.global(), pushedBackTag);
+
+      Iterable<WindowedValue<InputT>> pushedBackContents = pushedBack.read();
+      if (pushedBackContents != null) {
+        for (WindowedValue<InputT> elem : pushedBackContents) {
+
+          // we need to set the correct key in case the operator is
+          // a (keyed) window operator
+          setKeyContextElement1(new StreamRecord<>(elem));
+
+          pushbackDoFnRunner.processElement(elem);
+        }
+      }
+
+      setPushedBackWatermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+
+      pushbackDoFnRunner.finishBundle();
+
+      // maybe output a new watermark
```

Comment:
This code runs when we get the `+Inf` watermark on the side input(s). Above, we make sure to flush out all pushed back values.

This made me realise that we have to take into account the watermark hold here. The code should be 
```
// maybe output a new watermark
Instant watermarkHold = stateInternals.watermarkHold();

long potentialOutputWatermark = Math.min(currentInputWatermark, watermarkHold.getMillis());

if (potentialOutputWatermark > currentOutputWatermark) {
  currentOutputWatermark = potentialOutputWatermark;
  output.emitWatermark(new Watermark(currentOutputWatermark));
}```